src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -0,0 +1,354 @@
+pub const KEYBYTES: usize = $keybytes as usize;
+
+/// Number of bytes in a `Header`.
+/// An encrypted stream starts with a short header, whose size is HEADERBYTES
+/// bytes. That header must be sent/stored before the sequence of encrypted
+/// messages, as it is required to decrypt the stream.
+pub const HEADERBYTES: usize = $headerbytes as usize;
+
+/// Number of added bytes. The ciphertext length is guaranteed to always be
+/// message length + ABYTES.
+pub const ABYTES: usize = $abytes as usize;
+
+/// Tag message: the most common tag, that doesn't add any information about the
+/// nature of the message.
+const TAG_MESSAGE: u8 = $tag_message as u8;
+
+/// Tag push: indicates that the message marks the end of a set of messages, but
+/// not the end of the stream.
+/// For example, a huge JSON string sent as multiple chunks can use this tag to
+/// indicate to the application that the string is complete and that it can be
+/// decoded. But the stream itself is not closed, and more data may follow.
+const TAG_PUSH: u8 = $tag_push as u8;
+
+/// Tag rekey: "forget" the key used to encrypt this message and the previous
+/// ones, and derive a new secret key.
+const TAG_REKEY: u8 = $tag_rekey as u8;
+
+/// Tag final: indicates that the message marks the end of the stream and erases
+/// the secret key used to encrypt the previous sequence.
+const TAG_FINAL: u8 = $tag_final as u8;
+
+/// A tag is encrypted and attached to each message before the authentication
+/// code is generated over all data. A typical encrypted stream simply attaches
+/// `0` as a tag to all messages, except the last one which is tagged as
+/// `Tag::Final`. When decrypting the tag is retrieved and may be used.
+#[derive(Debug, PartialEq)]
+pub enum Tag {
+    /// Message, the most common tag, that doesn't add any information about the
+    /// nature of the message.
+    Message,
+    /// Push: indicates that the message marks the end of a set of messages, but
+    /// not the end of the stream.
+    /// For example, a huge JSON string sent as multiple chunks can use this tag
+    /// to indicate to the application that the string is complete and that it
+    /// can be decoded. But the stream itself is not closed, and more data may
+    /// follow.
+    Push,
+    /// Rekey: "forget" the key used to encrypt this message and the previous
+    /// ones, and derive a new secret key.
+    Rekey,
+    /// Final: indicates that the message marks the end of the stream and erases
+    /// the secret key used to encrypt the previous sequence.
+    Final,
+}
+
+impl Tag {
+    /// Returns the corresponding `Tag` given a `u8`, else `Err(())`.
+    fn from_u8(tag: u8) -> Result<Tag, ()> {
+        match tag {
+            TAG_MESSAGE => Ok(Tag::Message),
+            TAG_PUSH => Ok(Tag::Push),
+            TAG_REKEY => Ok(Tag::Rekey),
+            TAG_FINAL => Ok(Tag::Final),
+            _ => Err(())
+        }
+    }
+}
+
+new_type! {
+    /// `Key` for symmetric encryption.
```

Comment:
Symmetric authenticated encryption. I think it’s important to write it out consequently throughout the documentation. This is not an encryption primitive, it’s an authenticated encryption primitive (with authenticated additional data)

Reply:
Changed.